diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 18e1723..ce1a0fe 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -1619,7 +1619,7 @@ static unsigned int clientid_hashval(u32 id)
 
 static unsigned int clientstr_hashval(struct xdr_netobj name)
 {
-	return opaque_hashval(name.data, 8) & CLIENT_HASH_MASK;
+	return opaque_hashval(name.data, min(name.len, 8U)) & CLIENT_HASH_MASK;
 }
 
 /*
