diff --git a/drivers/net/can/usb/esd_usb.c b/drivers/net/can/usb/esd_usb.c
--- a/drivers/net/can/usb/esd_usb.c
+++ b/drivers/net/can/usb/esd_usb.c
@@ -1385,6 +1385,7 @@ static void esd_usb_disconnect(struct usb_interface *intf)

 	if (dev) {
 		dev->in_usb_disconnect = 1;
+		unlink_all_urbs(dev);
 		for (i = 0; i < dev->net_count; i++) {
 			if (dev->nets[i]) {
 				netdev = dev->nets[i]->netdev;
@@ -1393,7 +1394,6 @@ static void esd_usb_disconnect(struct usb_interface *intf)
 				free_candev(netdev);
 			}
 		}
-		unlink_all_urbs(dev);
 		kfree(dev);
 	}
 }
